simulator/src/main/java/kcl/teamIndexZero/traffic/simulator/data/trafficLight/TrafficLightController.java

```diff
@@ -1,17 +1,118 @@
+import java.time.format.DateTimeFormatter;
+import java.time.temporal.ChronoUnit;
+import java.util.ArrayList;
+import java.util.List;
+
 
 /**
  * Traffic Light controller
  * Controls the traffic lights in the map based on their rules
  */
 public class TrafficLightController implements ISimulationAware {
+        public String temp;
+        public long CurrentTime;
+        public LocalDateTime Temporary;
+        public long lastChange = 0;
+        public long timer = 5;
+        public long delay;
+        public static LocalDateTime start = LocalDateTime.of(1984, 12, 16, 7, 45, 55);
+        private TrafficLight model; //list this
+        private TrafficLightInSet modelSet; //list this
+        private TrafficLightSet set;
+        public List<TrafficLight> TrafficLightSinglesList;
+        public List<TrafficLightInSet> TrafficLightSetList;
+        SimulationTick simulationTick;
+        private static Logger_Interface LOG = Logger.getLoggerInstance(TrafficLightController.class.getSimpleName());
+
+       /**
+         * Constructor
+         */
+        public TrafficLightController(){
+            TrafficLightSinglesList = new ArrayList<TrafficLight>();
+            TrafficLightSetList = new ArrayList<TrafficLightInSet>();
+        }
+
+        public long formatTimeToLong(LocalDateTime date) {
+            long milliseconds = ChronoUnit.MILLIS.between(start, date);
+            return milliseconds;
+        }
+
+       /**
+         * Adds single traffic light to the List of all single traffic lights
+         *
+         * @param trafficLight object to be added to the list
+         */
+        public void addTrafficlight(TrafficLight trafficLight){
+
+            if (!(trafficLight == null)) {
+                TrafficLightSinglesList.add(trafficLight);
+                LOG.log("Added the following traffic lights to the set of all single Traffic Lights: ", trafficLight.getTrafficLightID() );
+            }
+            else {
+                LOG.log_Error("Error while adding to TrafficLightLnSet to the set");
+            }
+
+        }
+
+        /**
+         * Returns a list List of all single traffic lights
+         *
+         */
+         public List<TrafficLight> getSingleSet() {return TrafficLightSinglesList;}
+
+        /**
+         * {@inheritDoc}
+         */
+        @Override
+         public void tick(SimulationTick tick) {
+            CurrentTime = formatTimeToLong(simulationTick.getSimulatedTime());
+
+            /*when single traffic lights do not work in a synchronous way with the ticks*/
+            model.TrafficLightDelay = CurrentTime - lastChange;
+            if ((model.TrafficLightDelay >= timer) && (model.currentState == TrafficLightState.GREEN)) {
+                model.currentState = TrafficLightState.RED;
+            }
+
+            /*when set's traffic lights do not work in a synchronous way with the ticks*/
+            modelSet.TrafficLightInSetDelay = CurrentTime - lastChange;
```

Comment:
reference to `modelSet` and `model` but they are never instantiated. In anycase they are singular variable which gives us 1 TrafficLight and 1 TrafficLightInSet for every TrafficLightController. But we only want 1 controller. This is not going to work. Have a rethink.